File: Algorithms/309. Best Time to Buy and Sell Stock with Cooldown/swift/solution.swift
class Solution {
    func maxProfit(_ prices: [Int]) -> Int {
        var sold = Int.min, held = Int.min, reset = 0

        for price in prices {
            let preSold = sold

            sold = held + price
            held = max(held, reset - price)
            reset = max(reset, preSold)
        }

        return max(sold, reset)
    }
}

File: Algorithms/309. Best Time to Buy and Sell Stock with Cooldown/kotlin/solution.kt
class Solution {
    fun maxProfit(prices: IntArray): Int {
        var sold = Int.MIN_VALUE
        var held = Int.MIN_VALUE
        var reset = 0

        for (price in prices) {
            val preSold = sold
            sold = held + price
            held = maxOf(held, reset - price)
            reset = maxOf(reset, preSold)
        }

        return maxOf(sold, reset)
    }
}
